source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -1710,6 +1731,94 @@ namespace AdaptiveNamespace
         return S_OK;
     }
 
+    _Use_decl_annotations_ template<>
+    void XamlBuilder::SetAutoSize<IEllipse>(IEllipse* destination, IInspectable* parentElement, IInspectable* imageContainer, bool mustHideElement)
+    {
+        // Check if the image source fits in the parent container, if so, set the framework element's size to match the original image.
+        if (parentElement != nullptr && m_enableXamlImageHandling)
+        {
+            ComPtr<IInspectable> ellipseShape(imageContainer);
+            ComPtr<IShape> ellipseAsShape;
+            THROW_IF_FAILED(ellipseShape.As(&ellipseAsShape));
+
+            ComPtr<IBrush> ellipseBrush;
+            THROW_IF_FAILED(ellipseAsShape->get_Fill(&ellipseBrush));
+            ComPtr<IImageBrush> brushAsImageBrush;
+            THROW_IF_FAILED(ellipseBrush.As(&brushAsImageBrush));
+
+            ComPtr<IEllipse> ellipse(destination);
+
+            ComPtr<IUIElement> ellipseAsUIElement;
+            THROW_IF_FAILED(ellipse.As(&ellipseAsUIElement));
+
+            ComPtr<IImageSource> imageSource;
+            THROW_IF_FAILED(brushAsImageBrush->get_ImageSource(&imageSource));
+            ComPtr<IBitmapSource> imageSourceAsBitmap;
+            THROW_IF_FAILED(imageSource.As(&imageSourceAsBitmap));
+
+            // If the image hasn't loaded yet
+            if (mustHideElement)
+            {
+                // Collapse the Ellipse while the image loads, so that resizing is not noticeable
+                THROW_IF_FAILED(ellipseAsUIElement->put_Visibility(Visibility::Visibility_Collapsed));
+                // Handle ImageOpened event so we can check the imageSource's size to determine if it fits in its parent
+                EventRegistrationToken eventToken;
+                THROW_IF_FAILED(brushAsImageBrush->add_ImageOpened(
+                    Callback<IRoutedEventHandler>([ellipseAsUIElement](IInspectable* /*sender*/, IRoutedEventArgs * /*args*/) -> HRESULT {
+                    // Don't set the AutoImageSize on the ellipse as it makes the ellipse grow bigger than
+                    // what it would be otherwise, just set the visibility when we get the image
+                    return ellipseAsUIElement->put_Visibility(Visibility::Visibility_Visible);
+                })
+                    .Get(),
+                    &eventToken));
+            }
+        }
+    }
+
+    _Use_decl_annotations_ template<typename T>
```

Comment:
>_Use_decl_annotations_ [](start = 4, length = 22)

there aren't any declared annotations

Reply:
(also above)

---
In reply to: [236882414](https://github.com/Microsoft/AdaptiveCards/pull/2191#discussion_r236882414) [](ancestors = 236882414)